### PR TITLE
Derivative unit tests

### PR DESCRIPTION
**Pull request by ZedThree**
Basic sanity checks for all the first and second derivative kernels

Also a couple of useful helper utilities:

- `EnableIfField<T>` for enabling functions only if `T` is derived from `Field`
- `makeField<T>` returns a field with the result of calling a function at each point in index space

These are just in the unit tests at the mo, but could be moved into the main library if they're more generally useful

### CODE REVIEW

**Comment by d7919** [comment]
Is it worth adding the fourth order derivatives as well as these have the same signature as first/second?

File: tests/unit/include/test_derivs.cxx [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+// one more method than X and Y -- FFT. This means we can't just use
+// the provided `Combine` to produce the Cartesian product of
+// directions and methods as the latter depends on the
+// former. Instead, we instantiate the tests separately for each
+// direction and test all the methods for that direction in that
+// instantiation.
+
+namespace {
+// These are merely sanity checks, so don't expect them to agree very well!
+constexpr BoutReal derivatives_tolerance{1.e-2};
```
Comment (d7919) [inline tests/unit/include/test_derivs.cxx]: How small can we go currently?

File: tests/unit/include/test_derivs.cxx [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+#include <vector>
+
+// Some basic sanity checks for the derivative kernels. Checks the
+// derivatives of sin(R) where R = {X, Y, Z} for each R
+// individually. To make this as fast as possible, we use only a
+// couple of points in the non-tested directions -- not just one
+// though, as this allows us to check we're not introducing spurious
+// variation in the other directions.
+//
+// This is one of the more complicated uses of googletest! We need to
```
Comment (d7919) [inline tests/unit/include/test_derivs.cxx]: Missing word "need to..."

File: tests/unit/include/test_derivs.cxx [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+// These are merely sanity checks, so don't expect them to agree very well!
+constexpr BoutReal derivatives_tolerance{1.e-2};
+}
+
+class DerivativesTest
+    : public ::testing::TestWithParam<std::pair<DIRECTION, std::string>> {
+public:
+  DerivativesTest() : input{mesh}, first_order_expected{mesh} {
+
+    // Make sure fft functions are quiet by setting fft_measure to false
```
Comment (d7919) [inline tests/unit/include/test_derivs.cxx]: Has the side effect of making them deterministic.

File: tests/unit/include/test_derivs.cxx (hunk 1 above)
Comment (ZedThree) [reply tests/unit/include/test_derivs.cxx]: Smallest we can do for all the first derivatives is half this.

- `W2` fails below 5.e-3
- `C2` and `W3` fail below 2.e-4
- `S2` fails below 8.e-4
- `C4` fails below 5.e-7
- `FFT` fails below 1.e-14

We could pair the tolerances with the methods, but that would add additional complexity for what's supposed to be a basic test and I think would be more fragile.

I wouldn't be opposed to dropping the tolerance to 5.e-3, but I would be wary of doing so.

Comment (d7919) [reply tests/unit/include/test_derivs.cxx]: This is used as an absolute tolerance I think -- do you know what a relative tolerance limit would look like?

**Pull request merged**